### ISSUE TITLE
ENH: faster whitening matrix calculation for sm.GLS()

### ISSUE THREAD

**Issue by Wainberg** [statsmodels/statsmodels#8372]
`sm.GLS()` starts by calculating a whitening matrix, which the code calls `cholsigmainv`:

`cholsigmainv = np.linalg.cholesky(np.linalg.inv(sigma)).T`

But if you switch the order of operations so the Cholesky decomposition goes first (which is mathematically valid, if you get rid of the transpose), you only have to invert a lower triangular matrix instead of a dense matrix, which can be done efficiently with Lapack's [`dtrtri()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.lapack.dtrtri.html). 

`scipy.linalg.cholesky(..., lower=True)` also gives a slight speed-up over `np.linalg.cholesky(...)`, as does adding `overwrite_c=True` to `dtrtri()`. 

So I'd propose switching [this line](https://github.com/statsmodels/statsmodels/blob/edd843f0afb37b30cb503709e2d4ff4bc57f9c04/statsmodels/regression/linear_model.py#L181) to:

```
from scipy.linalg import cholesky
from scipy.linalg.lapack import dtrtri
cholsigmainv = dtrtri(cholesky(sigma, lower=True), lower=True, overwrite_c=True)[0]
```

This formulation is about four and a half times faster for a random 10,000 x 10,000 covariance matrix:

```
import numpy as np
from scipy.linalg import cholesky
from scipy.linalg.lapack import dtrtri
from timeit import timeit

arr = np.random.RandomState(0).random(size=(10000, 1000))
sigma = np.cov(arr) + 1e-8 * np.eye(len(arr))
timeit(lambda: np.linalg.cholesky(np.linalg.inv(sigma)).T, number=1)  # 304.3108371179551 seconds
timeit(lambda: dtrtri(cholesky(sigma, lower=True), lower=True, overwrite_c=True), number=1)  # 68.95121738314629 seconds
```

Calculating the whitening matrix becomes a major bottleneck for GLS when there are a lot of data points, and this enhancement should help overcome that.

**Comment by bashtage** [comment]
PR welcome - hopefully it will pass all existing tests.

**Comment by bashtage** [comment]
Does calling `dtrtri` require more exception handling by the user or in statsmodels, incase of a singular matrix? Or I suppose `cholesky` will ensure that it is non-singular?

**Comment by Wainberg** [comment]
Good point - `dtrtri` doesn't do the usual error handling that higher-level scipy functions do. Along with the inverse, it returns an integer `info`:
```
          = 0: successful exit
          < 0: if INFO = -i, the i-th argument had an illegal value
          > 0: if INFO = i, A(i,i) is exactly zero.  The triangular
               matrix is singular and its inverse can not be computed.
```
For the pull request, how should I handle info != 0? `raise LinAlgError('dtrtri returned non-zero exit code {}'.format(info))`?

**Comment by bashtage** [comment]
You should propogate this I think.  Ideally, the function would work the same after the PR.  You could do this.

1. Add a test with `pytest.raises` against the current mater using a singular `sigma`.  Make sure it raises and note the exception (LinAlgError?).  
2. Make the PR.
3. Ensure that the code still passes the test written in 1 with identical exception and message.

**Comment by Wainberg** [comment]
Okay, I made a pull request with the changes. I didn't add a test, it's got a PEP 8 issue, and I don't think the commit message is formatted according to statsmodels' conventions, but it should work. I ended up raising `LinAlgError('Singular matrix')` when it's singular (info > 0), which is the same as what `np.linalg.inv()` does. For info < 0, I raise a `ValueError` with a message that says the value of info.
